### ISSUE TITLE
2.01-22debian1+zfs3~wheezy postinst scripts fail to grub-install to mdraid

### ISSUE THREAD

**Issue by joshenders** [zfsonlinux/grub#7]
### Description

I recently performed an `apt-get upgrade` and was blindsided by this grub-pc package in the archive.zfsonlinux.org repo which rendered my systems using mdraid unbootable. I suspect the `DEBIAN/postinst` script in `2.01-22debian1+zfs3~wheezy` fails to install to mdraid sub-devices.

```
$ apt-cache policy grub-pc
grub-pc:
  Installed: 2.01-22debian1+zfs3~wheezy
  Candidate: 2.01-22debian1+zfs3~wheezy
  Version table:
 *** 2.01-22debian1+zfs3~wheezy 0
       1001 http://archive.zfsonlinux.org/debian/ wheezy/main amd64 Packages
        100 /var/lib/dpkg/status
     1.99-27+deb7u2 0
        500 http://mirrors.kernel.org/debian/ wheezy/main amd64 Packages
```
### Symptoms

After upgrading, I was greeted with a `grub rescue>` prompt. Although I could list the contents of my `(md/0)` device in the grub rescue shell, including the `/boot/grub` directory, I was unable to insmod `/boot/grub/i386-pc/normal.mod` and got a generic `error: file not found.` message. My only guess is that this is a stand-in message for binary incompatibility between `1.99-27+deb7u2` installed in the MBR of `/dev/sda` versus the `2.01-22debian1+zfs3~wheezy` modules on disk of `/dev/md0`.

I ended up using http://www.supergrubdisk.org/super-grub2-disk/ to boot into the system and performed a grub-install onto my mdraid sub-devices.

```
$ grub-install /dev/sda
$ grub-install /dev/sdb
```

I tested several iterations of this on multiple systems and it was unnecessary to `update-initramfs` or `update-grub`. Only the steps above.

**Comment by joshenders** [comment]
Doesn't look like the postinst script installed to `/dev/md0`:

```
$ dd if=/dev/md0 of=/tmp/mbr.bin bs=512 count=1
1+0 records in
1+0 records out
512 bytes (512 B) copied, 2.5816e-05 s, 19.8 MB/s
$ xxd /tmp/mbr.bin
0000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000060: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000070: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000080: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000090: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000100: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000140: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000150: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
0000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00001f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

**Comment by FransUrbo** [comment]
I'm not so sure this isn't unexpected.

When you first installed the machine, it asked you where/if to install grub. If you selected <code>/dev/md0</code> there, all is good - it wrote the loader in the place you specified...

When you then upgraded grub, it should have asked you (again) if you wanted to reinstall the grub boot loader. Did you not get such a question? I'm not sure if this is 'hidden' if the priority is to low...

Btw, if there's a missmatch between the grub boot loader and it's modules, you get another error than <code>error: file not found</code>.

**Comment by joshenders** [comment]
When I first installed the machine, we're talking about d-i now, I installed to `/dev/sda`. I went through a series of steps I outlined in this gist: https://gist.github.com/joshenders/3803617 to convert `/` to an mdraid volume. If any setting would be sticky, I would suspect it would be to `/dev/sda` since this was the original boot device. Notice in my last comment that `/dev/md0` still does not have the grub bootloader installed.

I was only prompted to merge my existing changes to `/etc/default/grub` for grub-pc but no grub-install type prompts. You can confirm this with the `/var/log/term.log` below. I'm familiar with debconf priority settings but I'm not manually overriding any default dpkg/apt settings.

Here's the prompt I get when I perform a `dpkg-reconfigure grub-pc`. Notice by default only `/dev/sdb` is checked. Perhaps this value is stored and upon upgrade the bootloader was only installed to `/dev/sdb`? If this is true, is there any reason why `1.99-27+deb7u2` bootloader code wouldn't boot updated `2.01-22debian1+zfs3~wheezy` userland?

![dpkg-reconfigure](http://cl.ly/image/0W2f1Z3N3L2N/Screen%20Shot%202014-02-24%20at%205.01.33%20PM.png)

Here's my `/var/log/term.log`:

```
Preparing to replace grub-pc 1.99-27+deb7u2 (using .../grub-pc_2.01-22debian1+zfs3~wheezy_amd64.deb) ...
Unpacking replacement grub-pc ...
Preparing to replace grub-pc-bin 1.99-27+deb7u2 (using .../grub-pc-bin_2.01-22debian1+zfs3~wheezy_amd64.deb) ...
Unpacking replacement grub-pc-bin ...
Preparing to replace grub2-common 1.99-27+deb7u2 (using .../grub2-common_2.01-22debian1+zfs3~wheezy_amd64.deb) ...
Unpacking replacement grub2-common ...
Preparing to replace grub-common 1.99-27+deb7u2 (using .../grub-common_2.01-22debian1+zfs3~wheezy_amd64.deb) ...
Unpacking replacement grub-common ...
...
Setting up grub-common (2.01-22debian1+zfs3~wheezy) ...
Installing new version of config file /etc/grub.d/00_header ...
Installing new version of config file /etc/grub.d/20_linux_xen ...
Installing new version of config file /etc/grub.d/41_custom ...
Installing new version of config file /etc/grub.d/05_debian_theme ...
Installing new version of config file /etc/grub.d/30_os-prober ...
Installing new version of config file /etc/grub.d/10_linux ...
Setting up grub2-common (2.01-22debian1+zfs3~wheezy) ...
Setting up grub-pc-bin (2.01-22debian1+zfs3~wheezy) ...
Setting up grub-pc (2.01-22debian1+zfs3~wheezy) ...
Installing new version of config file /etc/kernel/postinst.d/zz-update-grub ...
Installing new version of config file /etc/kernel/postrm.d/zz-update-grub ...
Package configuration┌───────────────────────────────────────────────────────┤ Configuring grub-pc ├───────────────────────────────────────────────────────┐│ A new version of configuration file /etc/default/grub is available, but the version installed currently has been locally modified.  │ │                                                                                                                                     │ │ What do you want to do about modified configuration file grub?                                                                      │ │                                                                                                                                     │ │                                     install the package maintainer's version                                                        │ │                                     keep the local version currently installed                                                      │ │                                     show the differences between the versions                                                       │ │                                     show a side-by-side difference between the versions                                             │ │                                     show a 3-way difference between available versions                                              │ │                                     do a 3-way merge between available versions (experimental)                                      │ │                                     start a new shell to examine the situation                                                      │ │                                                                                                                                     │ │                                                                                                                                     │ │                                                               <Ok>                                                                  │ │                                                                                                                                     │ └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                                                                                                                                        keep the local version currently installed                show the differences between the versions                 ^M^M
Package configuration┌───────────────────────────────┤ Configuring grub-pc ├────────────────────────────────┐│                                                                                      │ │ Line by line differences between versions                                            │ │                                                                                      │ │ --- /etc/default/grub 2014-02-16 04:40:47.786901538 -0800                            │ │ +++ /tmp/grub.iWIqjdmadw 2014-02-17 13:05:21.888533889 -0800                         │ │ @@ -9,16 +9,13 @@                                                                    │ │  GRUB_CMDLINE_LINUX_DEFAULT="quiet"                                                  │ │  GRUB_CMDLINE_LINUX="console=tty0 console=ttyS1,115200n8 ipv6.disable=1"             │ │                                                                                      │ │ -# - http://help.ovh.co.uk/IpmiSol                                                   │ │ -GRUB_SERIAL_COMMAND="serial --unit=1 --speed=115200 --word=8 --parity=no --stop=1"  │ │ -                                                                                    │ │  # Uncomment to enable BadRAM filtering, modify to suit your needs                   │ │  # This works with Linux (no patch required) and with any kernel that obtains        │ │  # the memory map information from GRUB (GNU Mach, kernel of FreeBSD ...)            │ │  #GRUB_BADRAM="0x01234567,0xfefefefe,0x89abcdef,0xefefefef"                          │ │                                                                                      │ │  # Uncomment to disable graphical terminal (grub-pc only)                            │ │ -GRUB_TERMINAL="console serial"                                                      │ │ +#GRUB_TERMINAL=console                                                              │ │                                                                                      │ │  # The resolution used on graphical terminal                                         │ │  # note that you can use only modes which your graphic card supports via VBE         │ │                                                                                      │ │                                        <Ok>                                          │ │                                                                                      │ └──────────────────────────────────────────────────────────────────────────────────────┘                                                                                         ^M^M
Package configuration┌───────────────────────────────────────────────────────┤ Configuring grub-pc ├───────────────────────────────────────────────────────┐│ A new version of configuration file /etc/default/grub is available, but the version installed currently has been locally modified.  │ │                                                                                                                                     │ │ What do you want to do about modified configuration file grub?                                                                      │ │                                                                                                                                     │ │                                     install the package maintainer's version                                                        │ │                                     keep the local version currently installed                                                      │ │                                     show the differences between the versions                                                       │ │                                     show a side-by-side difference between the versions                                             │ │                                     show a 3-way difference between available versions                                              │ │                                     do a 3-way merge between available versions (experimental)                                      │ │                                     start a new shell to examine the situation                                                      │ │                                                                                                                                     │ │                                                                                                                                     │ │                                                               <Ok>                                                                  │ │                                                                                                                                     │ └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                                                                                                                                        ^M^MInstallation finished. No error reported.
Generating grub.cfg ...
Found linux image: /boot/vmlinuz-3.2.0-4-amd64
Found initrd image: /boot/initrd.img-3.2.0-4-amd64
Found linux image: /boot/vmlinuz-3.2.0-3-amd64
Found initrd image: /boot/initrd.img-3.2.0-3-amd64
done
```

**Comment by joshenders** [comment]
I think there are actually two bugs here. One is the mdraid/grub-install integration/packaging issue but the other is that I never intended to update grub-pc to support ZFS in the first place. While a case could be made that the onus is on the user to be more vigilant during upgrades, it's rather bold of the zfsonlinux trust package to add a higher priority to the zfsonlinux origin if conflicting names are going to be used. Users are left to be hyper vigilant or at the mercy of ZoL. As grub-pc is very high-profile package, may I suggest renaming this package so that it doesn't clobber and its feature-set is more explicitly advertised?

```
$ dpkg -S /etc/apt/preferences.d/pin-zfsonlinux
zfsonlinux: /etc/apt/preferences.d/pin-zfsonlinux
$ cat /etc/apt/preferences.d/pin-zfsonlinux
cat /etc/apt/preferences.d/pin-zfsonlinux
Package: *
Pin: release o=archive.zfsonlinux.org
Pin-Priority: 1001
```

**Comment by FransUrbo** [comment]
> Generating grub.cfg ...
> Found linux image: /boot/vmlinuz-3.2.0-4-amd64
> Found initrd image: /boot/initrd.img-3.2.0-4-amd64
> Found linux image: /boot/vmlinuz-3.2.0-3-amd64
> Found initrd image: /boot/initrd.img-3.2.0-3-amd64
> done

Ok, so it did what it's supposed to when upgrading grub - update the config file, but not rewrite the boot block.

> One is the mdraid/grub-install integration/packaging issue

Which issue would that be?

> I never intended to update grub-pc to support ZFS in the first place.
> 
> If you add a repository to get packages from, you have to acknowledge the fact that there might be packages there that overwrites the one you have. 
> While a case could be made that the onus is on the user to be more vigilant during upgrades
> 
> That is always a good recommendation, wether or not you're using external/third party repositories or not, but especially true if you do... 

This is your "fault" (no accusation, just an observation) for not keeping tabs on what's going on, unfortunately.

> it's rather bold of the zfsonlinux trust package to add a higher priority to the zfsonlinux origin if conflicting names are going to be used.
> 
> Why? It IS a newer version! Are you suggesting that I'd call a 2.00 version something less than what it actually is, just so that people that doesn't want it, won't get it? Then what about those that DO want it? Those that added our repo (on purpose!) because they want the softwares and packages that we supply?

Sorry, but this is on you, not us.

> may I suggest renaming this package so that it doesn't clobber and its feature-set is more explicitly advertised?
> 
> I'm going to think about this, but "don't hold your breath". This will give all sorts of trouble for those that is using our grub on purpose... And I'm very unlikely to spend days to figure out and rename stuff for the very few that don't want the newer, ZFS enabled grub from the ZFS On Linux repository (the clue is in the name).

Now, to take this a step further - in Jessie (hopefully!), Debian GNU/Linux will provide ZoL packages in their repositories. I will most likely suspend my ZoL packages then, but if I don't (I probably continue with the dailies at least), should I name 'zfsutils" (etc) something else then just so that it won't clash? Call it 'zfsutils-zfs' maybe?! No, this won't happen. 

Those adding our repository will be considered asking for the latest versions of the packages we supply, without "funny names" and what not.

**Comment by joshenders** [comment]
> > Generating grub.cfg ...
> > Found linux image: /boot/vmlinuz-3.2.0-4-amd64
> > Found initrd image: /boot/initrd.img-3.2.0-4-amd64
> > Found linux image: /boot/vmlinuz-3.2.0-3-amd64
> > Found initrd image: /boot/initrd.img-3.2.0-3-amd64
> > done
> > Ok, so it did what it's supposed to when upgrading grub - update the config file, but not rewrite the boot block.

Correct. There was a boot block currently installed to `/dev/sda` as well as `/dev/sdb` from `1.99-27+deb7u2`. My question is, why did the system fail to boot with the `1.99-27+deb7u2` boot block and if that's a known issue, why didn't it prompt to install a new `2.01-22debian1+zfs3~wheezy` boot block?

> >  One is the mdraid/grub-install integration/packaging issue
> > Which issue would that be?

The above issue. I believe the correct behavior would have been to prompt to install an updated boot block. Which would be the screen below for `1.99-27+deb7u2`
![](http://cl.ly/image/0W2f1Z3N3L2N/Screen%20Shot%202014-02-24%20at%205.01.33%20PM.png)

> > I never intended to update grub-pc to support ZFS in the first place.
> > If you add a repository to get packages from, you have to acknowledge the fact that there might be  packages there that overwrites the one you have.

Forgive me if I'm wrong but that's not how I understand apt priorities to work when there is a competition between two sources for a package of the same name. If preference is given to external sources, ZoL would not need to pin their origin 1001 as they would have preference without doing so. Further review of APT_PREFERENCES(5) is needed to be certain.

```
$ cat /etc/apt/preferences.d/pin-zfsonlinux
cat /etc/apt/preferences.d/pin-zfsonlinux
Package: *
Pin: release o=archive.zfsonlinux.org
Pin-Priority: 1001
```

> >  While a case could be made that the onus is on the user to be more vigilant during upgrades
> > That is always a good recommendation, wether or not you're using external/third party repositories or not, but especially true if you do...
> 
> This is your "fault" (no accusation, just an observation) for not keeping tabs on what's going on, unfortunately.

Fully agree here.  My mistake, however I use many other third-party repositories, none of which pin their priority the way ZoL does. I believe this is unnecessary unless there are conflicting package names.

```
$ ls /etc/apt/sources.list.d/
archive.zfsonlinux.org.list  hwraid.le-vert.net.list
debian.saltstack.com.list    shell.ninthgate.se.list
download.videolan.org.list
```

> > it's rather bold of the zfsonlinux trust package to add a higher priority to the zfsonlinux origin if conflicting names are going to be used.
> > Why? It IS a newer version! Are you suggesting that I'd call a 2.00 version something less than what it actually is, just so that people that doesn't want it, won't get it? Then what about those that DO want it? Those that added our repo (on purpose!) because they want the softwares and packages that we supply?
> 
> Sorry, but this is on you, not us.

Unless the zfs patches to grub-pc have been merged upstream, I would consider ZoL packaging an alternate configuration or even a fork. In this case, it should probably be called something like grub-pc-zfs as it's not just a newer version, it's got an entirely different feature set. You'll have to clue me in here as I don't track the development as closely as I'm sure you do.

Similar patterns are seen with alternate configurations/featuresets of official Debian packages. One example that comes to mind is the bittorrent client deluge. In this example, we see the same program with several different packages for different feature sets and compile-time options:

```
$ apt-cache search deluge
deluge - bittorrent client written in Python/PyGTK
deluge-common - bittorrent client written in Python/PyGTK (common files)
deluge-console - bittorrent client written in Python/PyGTK (console ui)
deluge-gtk - bittorrent client written in Python/PyGTK (GTK+ ui)
deluge-torrent - bittorrent client (gtk ui transitional package)
deluge-web - bittorrent client written in Python/PyGTK (web ui)
deluge-webui - bittorrent client (web ui transitional package)
deluged - bittorrent client written in Python/PyGTK (daemon)
```

In the case the zfs patches are included in the latest version of grub-pc, it's kind of unusual that you would provide this package from archive.zfsonlinux.org and not just encourage users to install the latest version of grub-pc directly from testing. Not saying it's wrong, just a surprising convenience.

If archive.zfsonlinux.org is just providing a newer version of grub-pc which happens to support zfs, it may make sense to leave it named grub-pc until it becomes available in stable.

### Questions in summary:

Why did the system fail to boot with the `1.99-27+deb7u2` boot block or why didn't it prompt to re-install the `2.01-22debian1+zfs3~wheezy` boot block if there were incompatibilities?

If grub-pc offered in archive.zfsonlinux.org is a fork or adds new features not available in the latest grub-pc, will you consider renaming it so that it doesn't break our systems when we carelessly upgrade?

**Comment by FransUrbo** [comment]
> why did the system fail to boot with the 1.99-27+deb7u2 boot block
> 
> It didn't fail. You said yourself at the top, that you got the grub rescue> prompt.

That came from the boot block, but it could not continue after that, because everything else was 'wrong' - came from the 2.00 version.

> and if that's a known issue, why didn't it prompt to install a new 2.01-22debian1+zfs3~wheezy boot block?
> 
> If I don't misremembering, this is configurable somewhere - "always rewrite grub boot loader when upgrading" or something like that. You have to google it.
> Forgive me if I'm wrong but that's not how I understand apt priorities to work when there is a competition between two sources for a package of the same name.
> 
> Well, that is how it works. The latest version wins. No matter where it comes from.
> Unless the zfs patches to grub-pc have been merged upstream
> 
> It's not. And maybe never will. I've heard rumors that they're working on their own implementation, that doesn't need to link with ZFS... No idea how true that is or how far that's progressed. But if/when this comes to fruition, there won't be no need for grub in the ZoL repositories.
> it should probably be called something like grub-pc-zfs
> 
> This won't help in practice. It will need to Replace, Conflict and Provide grub-pc so that that won't be kept installed - it contain the same files as that package, which will stop grub-pc-zfs to install.

So grub-pc will have to be uninstalled anyway. So renaming it won't help...

**Comment by joshenders** [comment]
> > why did the system fail to boot with the 1.99-27+deb7u2 boot block
> 
> It didn't fail. You said yourself at the top, that you got the grub rescue> prompt.
> It "failed" in the sense that my machine didn't boot and required a non-trivial amount of effort to recover it after I installed `2.01-22debian1+zfs3~wheezy`. That is an issue that I would assume you care about fixing?
> 
> That came from the boot block, but it could not continue after that, because everything else was  'wrong' - came from the 2.00 version.

Earlier in this thread you mentioned that a different error is generated for that condition. (If `1.99-27+deb7u2` boot block and `2.01-22debian1+zfs3~wheezy` userspace are used together) 

> Btw, if there's a missmatch between the grub boot loader and it's modules, you get another error than error: file not found. 

I highly suspected this incompatibility between versions but glad to have confirmation. Let's talk about steps to take to fix this.

> > and if that's a known issue, why didn't it prompt to install a new 2.01-22debian1+zfs3~wheezy boot block?
> 
> If I don't misremembering, this is configurable somewhere - "always rewrite  grub boot loader when upgrading" or something like that. You have to google it.

I'll certainly look into that but let's be honest, if the maintainer can't remember the name of this obscure setting, what are the odds that a user will before an upgrade? `postinst` seems like the most appropriate place to prompt the user to reinstall their boot block as the alternative may be an unbootable system and hours of yak shaving.

> > Forgive me if I'm wrong but that's not how I understand apt priorities to work when there is a competition between two sources for a package of the same name.
> 
> Well, that is how it works. The latest version wins. No matter where it comes from.

I'm not describing a competition among versions, of course that's how we know apt to behave. What I'm describing is a competition among two sources providing the same package name at the same version. There is no preference, as far as I know, given to a source defined in `/etc/apt/sources.list.d/` vs `/etc/apt/sources.list` in which case, what is the need for pinning the archive.zfsonlinux.org origin as is done in the trust package? If only one repository provides a package by using a unique name, there is no competition to begin with and no need to pin.

> > Unless the zfs patches to grub-pc have been merged upstream
> 
> It's not. And maybe never will. I've heard rumors that they're working on their own implementation, that doesn't need to link with ZFS... No idea how true that is or how far that's progressed. But if/when this comes to fruition, there won't be no need for grub in the ZoL repositories.

That is by definition a fork, and should be named such. What you're doing now is maintaining a fork of grub-pc with zfs features and masquerading it as grub-pc.

> > it should probably be called something like grub-pc-zfs
> 
> This won't help in practice. It will need to Replace, Conflict and Provide grub-pc so that that won't be kept installed - it contain the same files as that package, which will stop grub-pc-zfs to install.
> 
> So grub-pc will have to be uninstalled anyway. So renaming it won't help...
> What you're describing sounds very much like the basic features provided by apt to provide conflict management and resolution. Can you elaborate on the challenges here?

What I understand you to be saying is that grub-pc and grub-pc-zfs are mutually exclusive, and proper Conflicts/Provides/Replaces lines in your control file are necessary for these to behave with one another. This seems like the most correct way to handle this? What is your reasoning for not doing this already?

**Comment by FransUrbo** [comment]
> Earlier in this thread you mentioned that a different error is generated for that condition.
> 
> Yes, but the grub modules are located in different places, which is the reason you got 'no such file'. If you change that (with a symlink or whatever), you'd get a completely different message.
> I'll certainly look into that but let's be honest, if the maintainer can't remember the name of this obscure setting, what are the odds that a user will before an upgrade?
> 
> I'm not the maintainer of grub. I'm at best the "recompiler for ZoL of grub". Which is different.
> what is the need for pinning the archive.zfsonlinux.org origin as is done in the trust package?
> 
> No idea. Never seen that before.

But that still doesn't matter! I'll tell you again, once last time (then I'm not even going to bother): 

There's a different (in every single way) between the packages that exist in Debian GNU/Linux and ZoL. This is intentional and on purpose. Anyone electing to add the ZoL repository, is going to be assumed to have chosen to get the packages we provide. Period, end of discussion and now lets never speak of it again.

**Comment by joshenders** [comment]
> > Earlier in this thread you mentioned that a different error is generated for that condition.
> 
> Yes, but the grub modules are located in different places, which is the reason you got 'no such file'. If you change that (with a symlink or whatever), you'd get a completely different message.

I see, that makes sense. This important detail was left out in your explanation.

> > I'll certainly look into that but let's be honest, if the maintainer can't remember the name of this obscure setting, what are the odds that a user will before an upgrade?
> 
> I'm not the maintainer of grub. I'm at best the "recompiler for ZoL of grub". Which is different.

The point is that you are much more familiar with grub-pc than an end user. 

> > what is the need for pinning the archive.zfsonlinux.org origin as is done in the trust package?
> 
> No idea. Never seen that before.

I've renamed mine, so it may be confusing. I'm talking about this file:

```
$ dpkg -L zfsonlinux | grep zfsonlinux.list
/etc/apt/sources.list.d/zfsonlinux.list
```

> But that still doesn't matter! I'll tell you again, once last time (then I'm not even going to bother):
> 
> There's a different (in every single way) between the packages that exist in Debian GNU/Linux and ZoL. This is intentional and on purpose. Anyone electing to add the ZoL repository, is going to be assumed to have chosen to get the packages we provide. Period, end of discussion and now lets never speak of it again.

What you're describing sounds a lot like, "we consider ZoL to be its own distribution" which I know is not the case as the release name in your `zfsonlinux.list` file is wheezy. The archive.zfsonlinux.org repo is clearly intended as a supplement to Wheezy.

```
$ cat /etc/apt/sources.list.d/archive.zfsonlinux.org.list
## This file is installed by the zfsonlinux package.

deb http://archive.zfsonlinux.org/debian wheezy main
```

### In summary:
1. You're maintaining a fork of grub-pc with zfs features that you are calling grub-pc in the archive.zfsonlinux.org repo.
2. You have no interest in renaming your fork of grub-pc or modifying the `control` file to play nicely with mainline wheezy grub-pc for reasons unknown.
3. ZoL users can expect to receive, "updates" to arbitrary packages as you see fit, "by design".
4. You aren't interested in fixing system breaking bugs introduced by this heavy handed approach.

I guess we're done here.

**Comment by FransUrbo** [comment]
> You aren't interested in fixing system breaking bugs introduced by this heavy handed approach.

That's a little harsh, but the answer to your other points is "yes".

I've tried to understand the 'original issue', but it's difficult because you have really tried to do your best to break your own system...

```
1. You started of by installing to sda
2. You then rebuilt this system to instead use md0
3. You have then done an upgrade of the system
```

This without "telling" your system that you've completely moved the installation to another location. Nor have you taken appropriate (whatever those should have been) action to make sure you're on top of the situation.

There IS something wrong with this in the sense that the upgrade from 1.99 to 2.01 only gave you a rescue prompt.

But the thing is, I can not reproduce this, nor can I understand why - I've done multiple installations of a bare-bones Wheezy install and then installed and upgrade to latest ZoL. So have others and no one have reported anything like this.

The only thing I have been able to come up with is that you moved your installation, and "everyone" got confused by this...
